File: BleModule/src/main/java/com/topdon/ble/ClassicScanner.java
package com.topdon.ble;

import android.bluetooth.BluetoothAdapter;

import androidx.annotation.NonNull;

import com.topdon.ble.util.Logger;

class ClassicScanner extends AbstractScanner {
    private boolean stopQuietly = false;

    ClassicScanner(EasyBLE easyBle, BluetoothAdapter bluetoothAdapter) {
        super(easyBle, bluetoothAdapter);
    }

    @Override
    protected boolean isReady() {
        return true;
    }

    @Override
    protected void performStartScan() {
        try {
            bluetoothAdapter.startDiscovery();
        } catch (SecurityException e) {
        }
    }

    @Override
    protected void performStopScan() {
        try {
            bluetoothAdapter.cancelDiscovery();
        } catch (SecurityException e) {
        }
    }

    @Override
    void setScanning(boolean scanning) {
        super.setScanning(scanning);
        if (scanning) {
            handleScanCallback(true, null, false, -1, "");
        } else if (!stopQuietly) {
            handleScanCallback(false, null, false, -1, "");
        } else {
            stopQuietly = false;
        }
    }

    @Override
    public void stopScan(boolean quietly) {
        if (isScanning()) {
            stopQuietly = quietly;
        }
        super.stopScan(quietly);
    }

    @NonNull
    @Override
    public ScannerType getType() {
        return ScannerType.CLASSIC;
    }
}

File: BleModule/src/main/java/com/mpdc4gsr/ble/ClassicScanner.kt
package com.mpdc4gsr.ble

import android.bluetooth.BluetoothAdapter
import android.content.Context
import android.util.Log
import com.mpdc4gsr.ble.callback.ScanListener
import com.mpdc4gsr.ble.util.BluetoothPermissionUtils

internal class ClassicScanner(easyBle: EasyBLE, bluetoothAdapter: BluetoothAdapter?) :
    AbstractScanner(easyBle, bluetoothAdapter) {
    private var stopQuietly = false

    override fun isReady(): Boolean {
        return true
    }

    override fun performStartScan() {
        val context: Context? = EasyBLE.Companion.getInstance().getContext()
        if (!BluetoothPermissionUtils.hasBluetoothScanPermission(context)) {
            Log.w(TAG, "Missing BLUETOOTH_SCAN permission for startDiscovery()")
            handleScanCallback(
                false, null, false, ScanListener.Companion.ERROR_LACK_BLUETOOTH_PERMISSION,
                "Missing Bluetooth scan permission"
            )
            return
        }

        try {
            bluetoothAdapter.startDiscovery()
        } catch (e: SecurityException) {
            Log.e(TAG, "SecurityException in startDiscovery(): " + e.message)
            handleScanCallback(
                false, null, false, ScanListener.Companion.ERROR_LACK_BLUETOOTH_PERMISSION,
                "Bluetooth permission denied: " + e.message
            )
        }
    }

    override fun performStopScan() {
        val context: Context? = EasyBLE.Companion.getInstance().getContext()
        if (!BluetoothPermissionUtils.hasBluetoothScanPermission(context)) {
            Log.w(TAG, "Missing BLUETOOTH_SCAN permission for cancelDiscovery()")
            return
        }

        try {
            bluetoothAdapter.cancelDiscovery()
        } catch (e: SecurityException) {
            Log.e(TAG, "SecurityException in cancelDiscovery(): " + e.message)
        }
    }

    override fun setScanning(scanning: Boolean) {
        super.setScanning(scanning)
        if (scanning) {
            handleScanCallback(true, null, false, -1, "")
        } else if (!stopQuietly) {
            handleScanCallback(false, null, false, -1, "")
        } else {
            stopQuietly = false
        }
    }

    override fun stopScan(quietly: Boolean) {
        if (isScanning()) {
            stopQuietly = quietly
        }
        super.stopScan(quietly)
    }

    override fun getType(): ScannerType {
        return ScannerType.CLASSIC
    }

    companion object {
        private const val TAG = "ClassicScanner"
    }
}
